Allow to specify ID when adding to the FAISS vectorstore.

atisharma opened the pull request:
# Allow to specify ID when adding to the FAISS vectorstore

This change allows unique IDs to be specified when adding documents / embeddings to a faiss vectorstore.

- This reflects the current approach with the chroma vectorstore.
- It allows rejection of inserts on duplicate IDs
- will allow deletion / update by searching on deterministic ID (such as a hash).
- If not specified, a random UUID is generated (as per previous behaviour, so non-breaking).

This commit fixes #5065 and #3896 and should fix #2699 indirectly. I've tested adding and merging.

Kindly tagging @Xmaster6y @dev2049 for review.

Commit:
Allow to specify ID when adding to the FAISS vectorstore.

This change allows unique IDs to be specified when adding documents /
embeddings to a faiss vectorstore.
This reflects the current approach with the chroma vectorstore. It
allows rejection of inserts on duplicate IDs and will allow deletion /
update by searching on deterministic ID (such as a hash).

This commit solves #5065 and #3896 and should solve #2699 indirectly.

Xmaster6y commented:
I think that all the `add` should have the extra id parameter.

Specifying the ids in the `from` class methods should also be possible.

atisharma commented:
Why?

They all pass through the `ids` argument to the underlying `__add` or `__from`.

Is it for the documentation?

Commit:
Add ids as optional arg to the public methods.

This gives visibility in the documentation.

atisharma commented:
I added the change you suggested.

Xmaster6y commented:
`__add` is not meant to be used outside the class as a "private method"

hwchase17 commented on langchain/vectorstores/faiss.py:
```diff
@@ -129,13 +132,15 @@ def add_texts(
         Args:
             texts: Iterable of strings to add to the vectorstore.
             metadatas: Optional list of metadatas associated with the texts.
+            ids: Optional list of unique IDs.
```
this should be passed to __add right?

atisharma replied:
It is, through the **kwargs.

atisharma replied:
done

Commit:
Fix passing the ids argument when calling __add, __from.

This is now necessary because they are listed in the input arguments
explicitly.

atisharma commented on langchain/vectorstores/faiss.py:
```diff
@@ -432,6 +443,7 @@ def from_embeddings(
         text_embeddings: List[Tuple[str, List[float]]],
         embedding: Embeddings,
         metadatas: Optional[List[dict]] = None,
+        ids: Optional[List[str]] = None,
```
done

atisharma commented on langchain/vectorstores/faiss.py:
```diff
@@ -161,6 +167,7 @@ def add_embeddings(
             text_embeddings: Iterable pairs of string and embedding to
                 add to the vectorstore.
             metadatas: Optional list of metadatas associated with the texts.
+            ids: Optional list of unique IDs.
```
done

atisharma commented:
> `__add` is not meant to be used outside the class as a "private method".

I know.

> I mean how do you specify the ids if you use add_texts?

This wasn't necessary when the ids were not explicitly listed as arguments for the public access methods, because the ids arguments would pass through to the private `__add` and `__from` methods through the `**kwargs`.

With the change you requested, yes, I should have passed them explicitly.

Commit:
Fix nasty typo.

Fix typo that would silently cause ids to be a sequential index instead
of the intended id, when using __from.

hwchase17 committed:
cr

hwchase17 committed:
cr

hwchase17 reviewed (approved):
thanks!
